lib/ABI/ModelAlignment.cpp

```diff
@@ -0,0 +1,172 @@
+    }
+    break;
+
+  case model::TypeKind::EnumType: {
+    // The alignment of an enum is the same as the alignment of its underlying
+    // type
+    const auto &Underlying = llvm::cast<model::EnumType>(&Type)->UnderlyingType;
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Underlying, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::TypedefType: {
+    // The alignment of an enum is the same as the alignment of its underlying
+    // type
+    const auto &Typedef = llvm::cast<model::TypedefType>(&Type)->UnderlyingType;
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Typedef, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::StructType: {
+    // The alignment of a struct is the same as the alignment of its most
+    // strictly aligned member.
+    const auto &Fields = llvm::cast<model::StructType>(&Type)->Fields;
+    if (auto A = rc_recur aggregateAlignmentImpl(VH, Fields, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::UnionType: {
+    // The alignment of a union is the same as the alignment of its most
+    // strictly aligned member.
+    const auto &Fields = llvm::cast<model::UnionType>(&Type)->Fields;
+    if (auto A = rc_recur aggregateAlignmentImpl(VH, Fields, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::Invalid:
+  case model::TypeKind::Count:
+  default:
+    rc_return std::nullopt;
+  }
+
+  VH.setSize(&Type, Alignment);
+
+  rc_return Alignment;
+}
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::QualifiedType &Type,
+                     const abi::Definition &ABI) {
+  // This code assumes that the QualifiedType is well formed.
+  for (auto It = Type.Qualifiers.begin(); It != Type.Qualifiers.end(); ++It) {
+    switch (It->Kind) {
+    case model::QualifierKind::Pointer:
+      // Doesn't matter what the type is, use alignment of the pointer.
+      rc_return ABI.TypeSpecific.at(It->Size).AlignAt;
+
+    case model::QualifierKind::Array: {
+      // The alignment of an array is the same as the alignment of its type.
+      const model::QualifiedType Element{
+        Type.UnqualifiedType, { std::next(It), Type.Qualifiers.end() }
+      };
+      if (auto MaybeAlignment = rc_recur naturalAlignmentImpl(VH, Element, ABI))
+        rc_return *MaybeAlignment;
+      else
+        rc_return std::nullopt;
+    }
+
+    case model::QualifierKind::Const:
+      // Const has no impact on alignment.
+      break;
+
+    case model::QualifierKind::Invalid:
+    case model::QualifierKind::Count:
+    default:
+      rc_return std::nullopt;
+    }
+  }
+
+  rc_return rc_recur naturalAlignmentImpl(VH, *Type.UnqualifiedType.get(), ABI);
+}
+
+RecursiveCoroutine<std::optional<uint64_t>>
```

Comment:
Is `naturalAlignment` recursive? If it's just `naturalAlignmentImpl` to be recursive, you can drop `RecursiveCoroutine`.

Reply:
Right `model::Type::size` does not need to be marked as a recursive coroutine. I'm already working on it on the inline primitives branch, but if you want you can fix it. Definitely let's not add new code that's unnecessarily a coroutine.

Reply:
As long as it's fully consistent with size, I'm up for any changes.

But now that you say it, all of this is going to be heavily affected by the inline primitives branch.

Then it's your call, since this will become your problem soon /s

Reply:
For now, just remove `RecursiveCoroutine<...>` from `naturalAlignment`, I'll fix the rest.

Reply:
What about the earlier related issues then? See the thread in mattermost.

Reply:
Done: https://github.com/revng/revng/pull/232/commits/12b1a8d64c9381087e2bbeefafda2e88a79e99cb and https://github.com/revng/revng/pull/232/commits/544081721e8bfa61e70661e316bfa247431ca965.